File: backend/middleware/validators.js
const { body, validationResult } = require('express-validator');

/**
 * Validation rules for user registration
 */
const validateRegister = [
  // Name validation
  body('name')
    .notEmpty().withMessage('Name is required')
    .isLength({ min: 2, max: 255 }).withMessage('Name must be between 2-255 characters')
    .matches(/^[a-zA-Z\s]+$/).withMessage('Name can only contain letters and spaces'),
  
  // Email validation
  body('email')
    .notEmpty().withMessage('Email is required')
    .isEmail().withMessage('Invalid email format')
    .isLength({ max: 255 }).withMessage('Email must be less than 255 characters'),
  
  // Password validation
  body('password')
    .notEmpty().withMessage('Password is required')
    .isLength({ min: 6 }).withMessage('Password must be at least 6 characters')
    .matches(/[a-z]/).withMessage('Password must contain at least one lowercase letter')
    .matches(/[A-Z]/).withMessage('Password must contain at least one uppercase letter')
    .matches(/\d/).withMessage('Password must contain at least one number'),
  
  // Phone validation
  body('phone')
    .notEmpty().withMessage('Phone number is required')
    .matches(/^\d{10}$/).withMessage('Phone must be exactly 10 digits'),
  
  // Address validation
  body('address')
    .notEmpty().withMessage('Address is required')
    .isLength({ min: 10, max: 500 }).withMessage('Address must be between 10-500 characters')
];

/**
 * Validation rules for user login
 */
const validateLogin = [
  // Email validation
  body('email')
    .notEmpty().withMessage('Email is required')
    .isEmail().withMessage('Invalid email format'),
  
  // Password validation
  body('password')
    .notEmpty().withMessage('Password is required')
    .isLength({ min: 6 }).withMessage('Password must be at least 6 characters')
];

/**
 * Middleware to check validation results
 * Must be used after validation rules
 */
const validate = (req, res, next) => {
  const errors = validationResult(req);
  
  if (!errors.isEmpty()) {
    console.log('Validation errors:', errors.array());
    
    // Format errors by field
    const formattedErrors = errors.array().map(error => ({
      field: error.param,
      message: error.msg
    }));
    
    return res.status(400).json({
      success: false,
      message: 'Validation failed',
      errors: formattedErrors
    });
  }
  
  next();
};

module.exports = { 
  validateRegister, 
  validateLogin,
  validate 
};

File: backend-python/utils/validators.py
import re
import logging

logger = logging.getLogger(__name__)

class ValidationError(Exception):
    """Custom validation exception"""
    pass

def validate_email(email):
    """Validate email format"""
    pattern = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
    if not re.match(pattern, email):
        raise ValidationError('Invalid email format (e.g., user@example.com)')
    if len(email) > 255:
        raise ValidationError('Email must be less than 255 characters')

def validate_password(password):
    """Validate password strength"""
    if len(password) < 6:
        raise ValidationError('Password must be at least 6 characters')
    if not re.search(r'[a-z]', password):
        raise ValidationError('Password must contain at least one lowercase letter')
    if not re.search(r'[A-Z]', password):
        raise ValidationError('Password must contain at least one uppercase letter')
    if not re.search(r'\d', password):
        raise ValidationError('Password must contain at least one number')

def validate_phone(phone):
    """Validate phone number"""
    if not re.match(r'^\d{10}$', phone):
        raise ValidationError('Phone must be exactly 10 digits (no spaces or dashes)')

def validate_name(name):
    """Validate name"""
    if len(name) < 2 or len(name) > 255:
        raise ValidationError('Name must be between 2-255 characters')
    if not re.match(r'^[a-zA-Z\s]+$', name):
        raise ValidationError('Name can only contain letters and spaces')

def validate_address(address):
    """Validate address"""
    if len(address) < 10 or len(address) > 500:
        raise ValidationError('Address must be between 10-500 characters')

def validate_issue_title(title):
    """Validate issue title"""
    if len(title) < 5 or len(title) > 200:
        raise ValidationError('Title must be between 5-200 characters')

def validate_issue_description(description):
    """Validate issue description"""
    if len(description) < 20 or len(description) > 2000:
        raise ValidationError('Description must be between 20-2000 characters')

def validate_priority(priority):
    """Validate priority"""
    valid_priorities = ['low', 'medium', 'high']
    if priority not in valid_priorities:
        raise ValidationError(f'Priority must be one of: {", ".join(valid_priorities)}')

def validate_status(status):
    """Validate status"""
    valid_statuses = ['open', 'in progress', 'resolved', 'closed', 'escalated']
    if status not in valid_statuses:
        raise ValidationError(f'Status must be one of: {", ".join(valid_statuses)}')

# ⭐ THIS WAS MISSING - ADDING NOW
def validate_all(email, password, name=None, phone=None, address=None):
    """Validate all fields and return errors if any"""
    errors = []
    
    # Validate email
    try:
        validate_email(email)
    except ValidationError as e:
        errors.append({'field': 'email', 'message': str(e)})
    
    # Validate password
    try:
        validate_password(password)
    except ValidationError as e:
        errors.append({'field': 'password', 'message': str(e)})
    
    # Optional fields for registration
    if name is not None:
        try:
            validate_name(name)
        except ValidationError as e:
            errors.append({'field': 'name', 'message': str(e)})
    
    if phone is not None:
        try:
            validate_phone(phone)
        except ValidationError as e:
            errors.append({'field': 'phone', 'message': str(e)})
    
    if address is not None:
        try:
            validate_address(address)
        except ValidationError as e:
            errors.append({'field': 'address', 'message': str(e)})
    
    return errors

